gabrielgomes94/leroy-merlin-test
# Feature/endpoints

### Feature

#### Spreadsheet Import

- [x] POST endpoint that receives an url with the direct link to the spreadsheet and imports it. 
- The endpoint creates a background job that will download the spreadsheet, then import it and save its product data on database. 

#### Endpoint to query spreadsheet status
- [x] Setup Laravel Job Status dependency to check the status of an import job.

- [x] Endpoint to check status of spreadsheet import

- [x] Change on Queue connection driver

- [x] Document endpoint

#### Endpoints to list, show, update and delete products
- [x] Implementation of the following endpoints for products:

  - index
  - update
  - show
  - delete

- [x] Handle errors and exceptions on API. 

- [x] Tests for these endpoints

- [x] Setup of test suit. 

### Setup

- Follow the instructions as describe on [readme.md](https://github.com/gabrielgomes94/leroy-merlin-test/pull/3/files#diff-0730bb7c2e8f9ea2438b52e419dd86c9) file

Observation: in case you have problems updating the database, you can recreate it for tests purposes:
```
php artisan migrate fresh
```

### Test

#### Import Spreadsheet

- Use the endpoint as described on [api.md](https://github.com/gabrielgomes94/leroy-merlin-test/pull/3/files#diff-00e71df22262087fd8ad820708997657). You can change the link to any other direct link to a spreadsheet. 
- Example spreadsheet link: https://s3-sa-east-1.amazonaws.com/teste-leroy-merlin/products_teste_webdev_leroy.xlsx

Another useful spreadsheet links:
- Spreadsheet with data errors: https://s3-sa-east-1.amazonaws.com/teste-leroy-merlin/products_teste_webdev_leroy_with_errors.xlsx

- Wrong spreadsheet(different format): https://s3-sa-east-1.amazonaws.com/teste-leroy-merlin/wrong_spreadsheet_teste_webdev_leroy.xlsx

- ODS spreadsheet with authentication error(S3 file is not public): https://s3-sa-east-1.amazonaws.com/teste-leroy-merlin/wrong_spreadsheet_teste_webdev_leroy.ods

#### Query an import status

1. Test `successful imports`:

- Start the queue: 
```
php artisan queue:listen database --tries=3
```
- Import a correct spreadsheet. 
Example: https://s3-sa-east-1.amazonaws.com/teste-leroy-merlin/products_teste_webdev_leroy.xlsx

- Test the status endpoint. It is expected to return a successful message.

2. Test `failure imports`:
- Start the queue: 
```
php artisan queue:listen database --tries=3
```
- Import wrong or invalid spreadsheets
Examples: 
  - https://s3-sa-east-1.amazonaws.com/teste-leroy-merlin/products_teste_webdev_leroy_with_errors.xlsx
  - https://s3-sa-east-1.amazonaws.com/teste-leroy-merlin/wrong_spreadsheet_teste_webdev_leroy.ods
  - https://s3-sa-east-1.amazonaws.com/teste-leroy-merlin/wrong_spreadsheet_teste_webdev_leroy.xlsx
- Test the status endpoint for these imports. It is expected to return a failure message and the error occurred. 

3. Test `pending imports`
- Stop the queue:
```
ctrl+c
```
- Import any spreadsheet. 
- Test the status endpoint for this import. It is expected to return a message saying the import is still being processed. 

#### Another Products CRUD endpoints:

Use the API as described on api.md file.
